chronos/asyncmacro2.nim

```diff
@@ -109,135 +108,157 @@ proc asyncSingleProc(prc: NimNode): NimNode {.compileTime.} =
+      #      against the spirit of the raises annotation - one should investigate
+      #      here the possibility of transporting more specific error types here
+      #      for example by casting exceptions coming out of `await`..
+      let raises = nnkBracket.newTree()
+      when not defined(chronosStrictException):
+        raises.add(newIdentNode("Exception"))
       else:
-        baseType
-    # Do not change this code to `quote do` version because `instantiationInfo`
-    # will be broken for `newFuture()` call.
-    outerProcBody.add(
-      newVarStmt(
-        retFutureSym,
-        newCall(newTree(nnkBracketExpr, ident "newFuture", subRetType),
-                newLit(prcName))
+        raises.add(newIdentNode("CatchableError"))
+        when (NimMajor, NimMinor) < (1, 4):
+          raises.add(newIdentNode("Defect"))
+
+      closureIterator.addPragma(nnkExprColonExpr.newTree(
+        newIdentNode("raises"),
+        raises
+      ))
+
+      # If proc has an explicit gcsafe pragma, we add it to iterator as well.
+      if prc.pragma.findChild(it.kind in {nnkSym, nnkIdent} and
+                              it.strVal == "gcsafe") != nil:
+        closureIterator.addPragma(newIdentNode("gcsafe"))
+      outerProcBody.add(closureIterator)
+
+      # -> var resultFuture = newFuture[T]()
+      # declared at the end to be sure that the closure
+      # doesn't reference it, avoid cyclic ref (#203)
+      var retFutureSym = ident "resultFuture"
+      var subRetType =
+        if returnType.kind == nnkEmpty:
+          newIdentNode("void")
+        else:
+          baseType
+      # Do not change this code to `quote do` version because `instantiationInfo`
+      # will be broken for `newFuture()` call.
+      outerProcBody.add(
+        newVarStmt(
+          retFutureSym,
+          newCall(newTree(nnkBracketExpr, ident "newFuture", subRetType),
+                  newLit(prcName))
+        )
+      )
+
+      # -> resultFuture.closure = iterator
+      outerProcBody.add(
+        newAssignment(
+          newDotExpr(retFutureSym, newIdentNode("closure")),
+          iteratorNameSym)
       )
-    )
- 
-    # -> resultFuture.closure = iterator
-    outerProcBody.add(
-       newAssignment(
-        newDotExpr(retFutureSym, newIdentNode("closure")),
-        iteratorNameSym)
-    )
-
-    # -> futureContinue(resultFuture))
-    outerProcBody.add(
-        newCall(newIdentNode("futureContinue"), retFutureSym)
-    )
-
-    # -> return resultFuture
-    outerProcBody.add newNimNode(nnkReturnStmt, prc.body[^1]).add(retFutureSym)
-
-  if prc.kind != nnkLambda: # TODO: Nim bug?
+
+      # -> futureContinue(resultFuture))
+      outerProcBody.add(
+          newCall(newIdentNode("futureContinue"), retFutureSym)
+      )
+
+      # -> return resultFuture
+      outerProcBody.add newNimNode(nnkReturnStmt, prc.body[^1]).add(retFutureSym)
+      outerProcBody
+    else:
+      nil
+  else:
+    nil
```

Comment:
Personally, I prefer to do
```nim
if not condition:
  nil
else:
  [long piece of code]
```
than what is being done here, because by the time we get to the `else`, we forgot the context of the original `if`. With the `if xx: nil else:` we have the full context in sight

Also fyi, https://github.com/status-im/nim-chronos/pull/251/files also added the ability to have async proc types

Reply:
removed the nils completely instead

> also added

it might be worth extracting any other bugfixes and refactoring from that branch so that we reduce the surface of merging it (and thus make it easier to revert should the need arise)

Reply:
(also, this PR fixes some incorrect raises, ie both Exception and CatchableError were being added redundantly)